File: clients/python/model/action.py
from model.building_action import BuildingAction
from model.move_action import MoveAction
from stream_wrapper import StreamWrapper
from typing import List

class Action:
    """TODO - Document"""

    __slots__ = ("moves","buildings",)

    moves: List[MoveAction]
    buildings: List[BuildingAction]

    def __init__(self, moves: List[MoveAction], buildings: List[BuildingAction]):
        self.moves = moves
        """TODO - Document"""
        self.buildings = buildings
        """TODO - Document"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "Action":
        """Read Action from input stream
        """
        moves = []
        for _ in range(stream.read_int()):
            moves_element = MoveAction.read_from(stream)
            moves.append(moves_element)
        buildings = []
        for _ in range(stream.read_int()):
            buildings_element = BuildingAction.read_from(stream)
            buildings.append(buildings_element)
        return Action(moves, buildings)
    
    def write_to(self, stream: StreamWrapper):
        """Write Action to output stream
        """
        stream.write_int(len(self.moves))
        for element in self.moves:
            element.write_to(stream)
        stream.write_int(len(self.buildings))
        for element in self.buildings:
            element.write_to(stream)
    
    def __repr__(self):
        return "Action(" + \
            repr(self.moves) + \
            ", " + \
            repr(self.buildings) + \
            ")"

File: clients/cpp/model/Action.cpp
#include "Action.hpp"

namespace model {

Action::Action(std::vector<model::MoveAction> moves, std::vector<model::BuildingAction> buildings) : moves(moves), buildings(buildings) { }

// Read Action from input stream
Action Action::readFrom(InputStream& stream) {
    std::vector<model::MoveAction> moves = std::vector<model::MoveAction>();
    size_t movesSize = stream.readInt();
    moves.reserve(movesSize);
    for (size_t movesIndex = 0; movesIndex < movesSize; movesIndex++) {
        model::MoveAction movesElement = model::MoveAction::readFrom(stream);
        moves.emplace_back(movesElement);
    }
    std::vector<model::BuildingAction> buildings = std::vector<model::BuildingAction>();
    size_t buildingsSize = stream.readInt();
    buildings.reserve(buildingsSize);
    for (size_t buildingsIndex = 0; buildingsIndex < buildingsSize; buildingsIndex++) {
        model::BuildingAction buildingsElement = model::BuildingAction::readFrom(stream);
        buildings.emplace_back(buildingsElement);
    }
    return Action(moves, buildings);
}

// Write Action to output stream
void Action::writeTo(OutputStream& stream) const {
    stream.write((int)(moves.size()));
    for (const model::MoveAction& movesElement : moves) {
        movesElement.writeTo(stream);
    }
    stream.write((int)(buildings.size()));
    for (const model::BuildingAction& buildingsElement : buildings) {
        buildingsElement.writeTo(stream);
    }
}

// Get string representation of Action
std::string Action::toString() const {
    std::stringstream ss;
    ss << "Action { ";
    ss << "moves: ";
    ss << "[ ";
    for (size_t movesIndex = 0; movesIndex < moves.size(); movesIndex++) {
        const model::MoveAction& movesElement = moves[movesIndex];
        if (movesIndex != 0) {
            ss << ", ";
        }
        ss << movesElement.toString();
    }
    ss << " ]";
    ss << ", ";
    ss << "buildings: ";
    ss << "[ ";
    for (size_t buildingsIndex = 0; buildingsIndex < buildings.size(); buildingsIndex++) {
        const model::BuildingAction& buildingsElement = buildings[buildingsIndex];
        if (buildingsIndex != 0) {
            ss << ", ";
        }
        ss << buildingsElement.toString();
    }
    ss << " ]";
    ss << " }";
    return ss.str();
}

}